sdks/java/core/src/main/java/org/apache/beam/sdk/io/Write.java

```diff
@@ -396,133 +470,133 @@ public void processElement(ProcessContext c) throws Exception {
     final PCollectionView<WriteOperation<T, WriteT>> writeOperationView =
         operationCollection.apply(View.<WriteOperation<T, WriteT>>asSingleton());
 
-    // Re-window the data into the global window and remove any existing triggers.
-    PCollection<T> inputInGlobalWindow =
-        input.apply(
-            Window.<T>into(new GlobalWindows())
-                .triggering(DefaultTrigger.of())
-                .discardingFiredPanes());
+    if (!windowedWrites) {
+      // Re-window the data into the global window and remove any existing triggers.
+      input =
+          input.apply(
+              Window.<T>into(new GlobalWindows())
+                  .triggering(DefaultTrigger.of())
+                  .discardingFiredPanes());
+    }
+
 
     // Perform the per-bundle writes as a ParDo on the input PCollection (with the WriteOperation
     // as a side input) and collect the results of the writes in a PCollection.
     // There is a dependency between this ParDo and the first (the WriteOperation PCollection
     // as a side input), so this will happen after the initial ParDo.
     PCollection<WriteT> results;
-    final PCollectionView<Integer> numShards;
-    if (computeNumShards == null) {
-      numShards = null;
-      results =
-          inputInGlobalWindow.apply(
-              "WriteBundles",
+    final PCollectionView<Integer> numShardsView;
```

Comment:
Somewhere in `expand` (if not compile time) we should check that the filename policy supports windowing OR the input is globally windowed.

And I guess part of the challenge is that `Write` doesn't know it has a `FileBasedSink`, so `Sink` will have to surface something. We can follow up by inlining them...